Parsing extension-av on Set Cookie header

ostorc opened the pull request:
This PR adds support for parsing extension-uv  and storing data retrieved from it.

Fixes #22077

ostorc committed:
Added Extension property to SetCookieHeaderValue and parsing for it

Tratcher commented on src/Http/Headers/src/SetCookieHeaderValue.cs:
```diff
@@ -554,12 +580,13 @@ public override bool Equals(object obj)
                 && SameSite == other.SameSite
-                && HttpOnly == other.HttpOnly;
+                && HttpOnly == other.HttpOnly
+                && HeaderUtilities.AreEqualCollections(Extensions, other.Extensions);
```
Does this consider order of extensions?

ostorc replied:
Changed that to consider order of extensions

Tratcher replied:
Sorry, I meant the opposite, order shouldn't matter.

ostorc replied:
Oh, I was thinking what you meant by that and chose the wrong option. I have reverted it back and changed the test for it.

Tratcher replied:
Yeah, I just meant to ask about the behavior of HeaderUtilities.AreEqualCollections.

ostorc replied:
Just to confirm things. The HeaderUtilities.AreEqualCollections is indeed checking if two unordered list have same content. It does it ing O(n*m) time, but since the number of extensions is not likely to reach any high number, so it doesn't really matter.

Tratcher reviewed (commented):
Seems like a good start. The tests will be the most telling for a change like this. Don't forget to update the reference assemblies.

ostorc committed:
Update reference assemblies

ostorc committed:
Change Equals() to respect order of Extensions

ostorc committed:
Added test data for parsing of extension-av

ostorc committed:
Fix of parsing multiple cookies in one string

Because extension-av doesn't have any fixed format we have to look after
commas when parsing to determine end of Set-Cookie string

ostorc committed:
Extension-av order doesn't matter

Tratcher reviewed (commented):
A few style issues to address.

The commas are my biggest behavioral question though.

BrennanConroy commented:
@ostorc Can you take a look at the feedback?

shirhatti commented:
@ostorc Are you still working on this?

ostorc commented:
Hi, I am sorry for the dealy. I had really busy week. I will finish changes today and update the PR.

ostorc committed:
Updated SetCookieHeaderValue to reflect comments in PR

ostorc committed:
Update ref

ostorc committed:
Extend tests. Mainly check all possible combinations of SetCookieHeaders

ostorc commented on src/Http/Headers/ref/Microsoft.Net.Http.Headers.netcoreapp.cs:
```diff
@@ -329,6 +329,7 @@ public partial class SetCookieHeaderValue
         public SetCookieHeaderValue(Microsoft.Extensions.Primitives.StringSegment name, Microsoft.Extensions.Primitives.StringSegment value) { }
         public Microsoft.Extensions.Primitives.StringSegment Domain { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute] set { } }
         public System.DateTimeOffset? Expires { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute] set { } }
+        public System.Collections.Generic.List<Microsoft.Extensions.Primitives.StringSegment> Extensions { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute] set { } }
```
Fixed

ostorc commented on src/Http/Headers/src/SetCookieHeaderValue.cs:
```diff
@@ -498,13 +525,13 @@ private static int GetSetCookieLength(StringSegment input, int startIndex, out S
+                    ReadToSemicolonOrEnd(input, ref offset, includeComma: true);
+
+                    if (result.Extensions == null)
+                        result.Extensions = new List<StringSegment>();
```
Updated all cases

ostorc committed:
Comment of special case

ostorc commented on src/Http/Headers/src/SetCookieHeaderValue.cs:
```diff
@@ -523,14 +550,28 @@ private static bool ReadEqualsSign(StringSegment input, ref int offset)
         }
 
-        private static StringSegment ReadToSemicolonOrEnd(StringSegment input, ref int offset)
+        private static StringSegment ReadToSemicolonOrEnd(StringSegment input, ref int offset, bool includeComma = false)
```
Changed default value to check comma too. Added new test data, which check all combinations of valid data and the test showed that this is needed change. Only exception is parsing of Expire, where in date may be comma.

ostorc commented on src/Http/Headers/src/SetCookieHeaderValue.cs:
```diff
@@ -98,6 +99,8 @@ public StringSegment Value
 
         public bool HttpOnly { get; set; }
 
+        public List<StringSegment> Extensions { get; set; }
```
Fixed that to:
```csharp
 public IList<StringSegment> Extensions { get; } = new List<StringSegment>();
```

Tratcher reviewed (approved):
Sorry, one merge conflict to resolve and I think you're done.

ostorc committed:
Merge branch 'master' into master

ostorc commented:
Conflict resolved

Tratcher commented:
Thanks
